man/bcat.1
.\" generated with Ronn/v0.6.42
.\" http://github.com/rtomayko/ronn/tree/0.6.6-36-gb67d494
.
.TH "BCAT" "1" "June 2010" "Ryan Tomayko" "Bcat 0.5.0"
.
.SH "NAME"
\fBbcat\fR \- browser cat
.
.SH "SYNOPSIS"
\fBbcat\fR [\-htad] [\fIfile\fR\.\.\.]
.
.br
\fBbtee\fR [\-htad] [\fIfile\fR\.\.\.]
.
.SH "DESCRIPTION"
The \fBbcat\fR utility reads from standard input, or one or more \fIfile\fRs, and pipes output into a web browser\. \fIfile\fR may be \'\-\', in which case standard input is concatenated at that position\.
.
.P
When invoked as \fBbtee\fR, all input is written immediately to standard output in addition to being piped into the browser\.
.
.SH "OPTIONS"
\fBbcat\fR opens a simple, undecorated page with the system default web browser and immediately begins streaming input\. The following options control the browser display:
.
.TP
\fB\-b\fR, \fB\-\-browser\fR=default|firefox|safari|chrome|opera|\fIother\fR
The name of the browser application\. Defaults to the value of the \fBBCAT_BROWSER\fR environment variable, or the system default browser when no \fBBCAT_BROWSER\fR is defined\.
.
.TP
\fB\-T\fR, \fB\-\-title\fR=\fItext\fR
Use \fItext\fR as the page \fB<title>\fR\. By default, the path to the current working directory is used as the title\.
.
.TP
\fB\-a\fR, \fB\-\-ansi\fR
Turns on VT100/ANSI escape sequence conversion\. This causes all input to be piped through a2h(1), replacing ANSI escape sequences with HTML for things like bold, underline, and colors\. On by default when the input is text; use the \fB\-a\fR option to turn it on when the input is HTML\.
.
.P
By default, \fBbcat\fR attempts to detect whether input is HTML or plain text using a simple heuristic, but you can force input to be treated as one or the other with these options:
.
.TP
\fB\-t\fR, \fB\-\-text\fR
The input is non\-HTML encoded text\. All bare \fB<\fR and \fB&\fR characters are entity encoded, end\-of\-line characters are converted to \fB<br>\fR, and the entire output is wrapped in a \fB<pre>\fR\.
.
.TP
\fB\-h\fR, \fB\-\-html\fR
The input is already HTML encoded\. Under this mode, bcat passes input through to the browser mostly unmodified\. The input may be a full HTML document, or it may be an HTML fragment\. \fBbcat\fR outputs \fB<html>\fR, \fB<head>\fR, and \fB<body>\fR elements even if they are not included in the input\.
.
.P
Miscellaneous options:
.
.TP
\fB\-d\fR, \fB\-\-debug\fR
Turn on verbose debug logging to standard error\. Include this output when reporting defects if applicable\.
.
.SH "ENVIRONMENT"
.
.TP
\fBBCAT_BROWSER\fR=default|firefox|safari|chrome|opera|\fIother\fR
The name of the browser to use by default\. \fBbcat\fR maps this to an actual browser command based on the current platform\. The special value \"default\" maps to the system default browser\.
.
.TP
\fBBCAT_COMMAND\fR=\fIcommand\fR
The entire browser command line (to be executed by \fB/bin/sh\fR)\. This overrides the \fBBCAT_BROWSER\fR environment variable and makes the \fB\-\-browser\fR (\fB\-b\fR) option a no\-op\. This should only be necessary when running a browser unknown to bcat or in order to pass special arguments\.
.
.SH "EXAMPLES"
.
.SS "Basic Shell Examples"
With build tools:
.
.IP "" 4
.
.nf

make test |bcat
rake test |bcat
.
.fi
.
.IP "" 0
.
.P
As a clipboard viewer:
.
.IP "" 4
.
.nf

pbpaste  |bcat  # mac
xclip \-o |bcat  # X11
.
.fi
.
.IP "" 0
.
.P
For previewing HTML:
.
.IP "" 4
.
.nf

markdown README\.md |bcat
redcloth README\.textile |bcat
erb \-T \- template\.erb |bcat
mustache < template\.mustache |bcat
pygmentize \-Ofull,style=colorful \-f html main\.c |bcat
.
.fi
.
.IP "" 0
.
.P
As a simple man pager:
.
.IP "" 4
.
.nf

export MANPAGER=\'col \-b |bcat\'
man grep
.
.fi
.
.IP "" 0
.
.P
With git, selectively:
.
.IP "" 4
.
.nf

git log \-p \-\-color |bcat
git diff \-\-color HEAD@{5d} HEAD |bcat
.
.fi
.
.IP "" 0
.
.P
With git, as the default PAGER:
.
.IP "" 4
.
.nf

export GIT_PAGER=bcat

git log \-p
git diff HEAD@{5d} HEAD
.
.fi
.
.IP "" 0
.
.P
As a log viewer:
.
.IP "" 4
.
.nf

tail \-n 1000 \-f log/development\.log |bcat
tail \-f $RAILS_ROOT/log/development\.log
.
.fi
.
.IP "" 0
.
.P
Or, a remote log viewer:
.
.IP "" 4
.
.nf

ssh example\.org \'tail \-n 1000 \-f /var/log/syslog\' |bcat
.
.fi
.
.IP "" 0
.
.SS "Vim and vi Examples"
Preview current buffer as HTML:
.
.IP "" 4
.
.nf

:!markdown % |bcat
:!ronn \-5 \-\-pipe % |bcat
.
.fi
.
.IP "" 0
.
.P
Create keymappings:
.
.IP "" 4
.
.nf

map ,pm :!markdown % |bcat
map ,pp :!pygmentize \-Ofull,style=colorful \-f html % |bcat
.
.fi
.
.IP "" 0
.
.P
Use with \fBmakeprg\fR:
.
.IP "" 4
.
.nf

set makeprg=make\e \e\e\e|bcat
set makeprg=markdown\e %\e \e\e\e|bcat
set makeprg=testrb\e %\e \e\e\e|bcat
.
.fi
.
.IP "" 0
.
.SH "SEE ALSO"
cat(1), tee(1), open(1)


lib/bcat.rb
require 'rack'
require 'bcat/reader'
require 'bcat/ansi'
require 'bcat/html'
require 'bcat/server'
require 'bcat/browser'

class Bcat
  VERSION = '0.5.0'
  include Rack::Utils

  attr_reader :format

  def initialize(files=[], config={})
    @config = {:Host => '127.0.0.1', :Port => 8091}.merge(config)
    @reader = Bcat::Reader.new(files)
    @format = @config[:format]
  end

  def [](key)
    @config[key]
  end

  def to_app
    app = self
    Rack::Builder.new do
      use Rack::Chunked
      run app
    end
  end

  def serve!(&bk)
    Bcat::Server.run to_app, @config, &bk
  end

  def call(env)
    notice "#{env['REQUEST_METHOD']} #{env['PATH_INFO'].inspect}"
    [200, {"Content-Type" => "text/html;charset=utf-8"}, self]
  end

  def assemble
    @format = @reader.sniff if @format.nil?

    @filter = @reader
    @filter = TeeFilter.new(@filter) if @config[:tee]
    @filter = TextFilter.new(@filter) if @format == 'text'
    @filter = ANSI.new(@filter) if @format == 'text' || @config[:ansi]
  end

  def each
    assemble

    head_parser = Bcat::HeadParser.new

    @filter.each do |buf|
      if head_parser.nil?
        yield buf
      elsif head_parser.feed(buf)
        yield content_for_head(inject=head_parser.head)
        yield head_parser.body
        head_parser = nil
      end
    end

    if head_parser
      yield content_for_head(inject=head_parser.head) +
            head_parser.body
    end

    yield foot
  rescue Errno::EINVAL
    # socket was closed
    notice "browser client went away"
  rescue => boom
    notice "boom: #{boom.class}: #{boom.to_s}"
    raise
  end

  def content_for_head(inject='')
    [
      "\n" * 1000,
      "<!DOCTYPE html>",
      "<html>",
      "<head>",
      "<!-- bcat was here -->",
      "<title>#{self[:title] || 'bcat'}</title>",
      inject.to_s,
      "</head>"
    ].join("\n")
  end

  def foot
    "</body></html>"
  end

  def escape_js(string)
    string = string.gsub(/['\\]/) { |char| "\\#{char}" }
    string.gsub!(/\n/, '\n')
    string
  end

  def close
    notice "closing with interrupt"
    raise Interrupt
  end

  def notice(message)
    return if !@config[:debug]
    warn "#{File.basename($0)}: #{message}"
  end
end
